portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```
@@ -58,33 +67,38 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String newComment = request.getParameter("comment");
-    if (newComment != null && !newComment.isEmpty()) {
+    String message = request.getParameter("comment");
+    long parent = getCommentParent(request);
+    if (message != null && !message.isEmpty()) {
       // TODO: Pass in actual values for parent and score.
-      Entity commentEntity = Comment.createComment(newComment, 0, 0);
+      Entity commentEntity = Comment.createComment(message, parent, 0);
       ds.put(commentEntity);
     }
 
     response.sendRedirect(getRedirect(request));
   }
 
-  private String getRedirect(HttpServletRequest request) {
-    String referer = request.getHeader("referer");
-    String referer_host;
+  private long getCommentParent(HttpServletRequest request) {
+    String parentString = request.getParameter("parent");
 
-    // It's possible the user manually set a referer that is not a valid URI
+    long parent;
     try {
-      referer_host = new URI(referer).getHost();
-    } catch (URISyntaxException e) {
-      referer_host = "";
+      parent = Long.parseLong(parentString);
+    } catch (NumberFormatException e) {
+      System.err.println("Could not convert '" + parentString + "' to long.");
+      // Make it a top-level comment if we can't parse the parent,
+      parent = 0;
```

I think it would be better to let it fail (throw an exception) if the parent id is not valid instead of returning the default value. It would be a weird experience if a user added a reply but it silently becomes a top level comment. (Same for the below parent < 0 block)

---

+1 for throwing an exception in this case, as well as the case where parent < 0.